src/main/host/syscall/unistd.rs

```diff
@@ -0,0 +1,316 @@
+    let fd_ptr = unsafe { args.args[0].as_ptr };
+    let flags = unsafe { args.args[1].as_u64 } as libc::c_int;
+
+    pipe_helper(sys, fd_ptr, flags)
+}
+
+fn pipe_helper(sys: &mut c::SysCallHandler, fd_ptr: c::PluginPtr, flags: i32) -> c::SysCallReturn {
+    // make sure they didn't pass a NULL pointer
+    if fd_ptr.val == 0 {
+        return c::SysCallReturn::from_errno(nix::errno::Errno::EFAULT);
+    }
+
+    // pipe flags that we support
+    let supported_flags = libc::O_NONBLOCK & libc::O_CLOEXEC;
+    let flags_to_set = flags & supported_flags;
+
+    // the user requested flags that we don't support
+    if flags != flags_to_set {
+        // exit early if the O_DIRECT flag was set
+        if flags & libc::O_DIRECT != 0 {
+            warn!("We don't currently support pipes in 'O_DIRECT' mode");
+            return c::SysCallReturn::from_errno(nix::errno::Errno::EOPNOTSUPP);
+        }
+        warn!("Ignoring pipe flags");
+    }
+
+    // reference-counted buffer for the pipe
+    let buffer = pipe::SharedBuf::new();
+    let buffer = Arc::new(AtomicRefCell::new(buffer));
+
+    // reference-counted file object for read end of the pipe
+    let reader = pipe::PipeFile::new(Arc::clone(&buffer), FileMode::READ);
+    let reader = Arc::new(AtomicRefCell::new(PosixFile::Pipe(reader)));
+
+    // reference-counted file object for write end of the pipe
+    let writer = pipe::PipeFile::new(Arc::clone(&buffer), FileMode::WRITE);
+    let writer = Arc::new(AtomicRefCell::new(PosixFile::Pipe(writer)));
+
+    // set the file objects to listen for events on the buffer
+    pipe::PipeFile::enable_notifications(&reader);
+    pipe::PipeFile::enable_notifications(&writer);
+
+    // file descriptors for the read and write file objects
+    let mut reader_desc = Descriptor::new(reader);
+    let mut writer_desc = Descriptor::new(writer);
+
+    // set the file descriptor flags
+    reader_desc.set_flags(flags_to_set);
+    writer_desc.set_flags(flags_to_set);
+
+    // register the file descriptors and return them to the caller
+    let num_items = 2;
+    let size_needed = num_items * std::mem::size_of::<libc::c_int>();
+    let fd_ptr =
+        unsafe { c::process_getWriteablePtr(sys.process, sys.thread, fd_ptr, size_needed as u64) };
+    let fds = unsafe { std::slice::from_raw_parts_mut(fd_ptr as *mut libc::c_int, num_items) };
+
+    fds[0] = unsafe {
+        c::process_registerCompatDescriptor(
+            sys.process,
+            Box::into_raw(Box::new(CompatDescriptor::New(reader_desc))),
+        )
+    };
+    fds[1] = unsafe {
+        c::process_registerCompatDescriptor(
+            sys.process,
+            Box::into_raw(Box::new(CompatDescriptor::New(writer_desc))),
+        )
+    };
+
+    debug!("Created pipe reader fd {} and writer fd {}", fds[0], fds[1]);
+
+    c::SysCallReturn::from_int(0)
+}
+
+mod export {
+    use super::*;
+
+    #[no_mangle]
```

Comment:
Move to module-level?

Reply:
As above, tried but wasn't able to get it to work.